contracts/test/PlonkVerifier.t.sol

```diff
@@ -0,0 +1,432 @@
+        cmds[1] = "plonk-batch-verify";
+        cmds[2] = vm.toString(uint32(1));
+
+        bytes memory result = vm.ffi(cmds);
+        (
+            IPlonkVerifier.VerifyingKey[] memory verifyingKeys,
+            uint256[][] memory publicInputs,
+            IPlonkVerifier.PlonkProof[] memory proofs,
+            bytes[] memory extraTranscriptInitMsgs
+        ) = abi.decode(
+            result,
+            (IPlonkVerifier.VerifyingKey[], uint256[][], IPlonkVerifier.PlonkProof[], bytes[])
+        );
+
+        proofs[0] = badProof;
+        assert(!V.batchVerify(verifyingKeys, publicInputs, proofs, extraTranscriptInitMsgs));
+    }
+
+    /// forge-config: default.fuzz.runs = 30
+    /// @dev Test when bad extraTranscriptInitMsgs are supplied, the verification should fail
+    function testFuzz_badExtraTranscriptInitMsgs_fails(bytes calldata badMsg) external {
+        string[] memory cmds = new string[](3);
+        cmds[0] = "diff-test";
+        cmds[1] = "plonk-batch-verify";
+        cmds[2] = vm.toString(uint32(1));
+
+        bytes memory result = vm.ffi(cmds);
+        (
+            IPlonkVerifier.VerifyingKey[] memory verifyingKeys,
+            uint256[][] memory publicInputs,
+            IPlonkVerifier.PlonkProof[] memory proofs,
+            bytes[] memory extraTranscriptInitMsgs
+        ) = abi.decode(
+            result,
+            (IPlonkVerifier.VerifyingKey[], uint256[][], IPlonkVerifier.PlonkProof[], bytes[])
+        );
+
+        extraTranscriptInitMsgs[0] = badMsg;
+        assert(!V.batchVerify(verifyingKeys, publicInputs, proofs, extraTranscriptInitMsgs));
+    }
+}
+
+contract PlonkVerifier_validateProof_Test is PlonkVerifierCommonTest {
+    /// @dev Test that a valid proof shouldn't revert
+    function test_validProof_succeeds() external {
+        // a valid proof
+        IPlonkVerifier.PlonkProof memory proof = dummyProof(42);
+
+        V._validateProof(proof);
+    }
+
+    /// forge-config: default.fuzz.runs = 30
+    /// @dev Randomly pick a coordinate of a point among points in a proof
+    /// mutate it to another value so that the point is no longer valid,
+    /// test if our check will revert.
+    function testFuzz_RevertIfProofContainsInvalidGroup(uint256 nthPoint, bool testX) external {
+        // a valid proof
+        IPlonkVerifier.PlonkProof memory proof = dummyProof(42);
+
+        // we are testing the `nthPoint` in the `proof`,
+        // only mutating a single field element (either x or y coordinate)
+        // There are 13 points in total.
+        nthPoint = bound(nthPoint, 0, 12);
+
+        assembly {
+            if testX {
+                // muteate the x coordinate
+                mstore(mload(add(proof, mul(0x20, nthPoint))), 0x1234)
+            }
+            // else, mutate y coordinate
+            mstore(add(mload(add(proof, mul(0x20, nthPoint))), 0x20), 0x1234)
+        }
+
+        vm.expectRevert();
+        V._validateProof(proof);
```

Comment:
How do we know the proof validation failed because the group element is invalid and not because the proof is incorrect w.r.t. to the public inputs / verification key?

Reply:
wait... what do you mean?
we are testing `_validateProof` function, which only checks group & field elements validity, not doing anything else. (it doesn't even accept public inputs or vk)

Reply:
Ah sorry, I get it now. Thanks.